tests/unit/cluster/atomic-slot-migration.tcl

```diff
@@ -99,29 +99,54 @@ proc populate_slot {num args} {
     R $idx deferred 0
 }
 
+# Return 1 if all instances are idle
+proc asm_all_instances_idle {total} {
+    for {set i 0} {$i < $total} {incr i} {
+        if {[CI $i slot_migration_active_tasks] != 0} { return 0 }
+        if {[CI $i slot_migration_active_trim_running] != 0} { return 0 }
+    }
+    return 1
+}
+
 # Wait for all ASM tasks to complete in the cluster
 proc wait_for_asm_done {} {
     set total_instances [expr {$::cluster_master_nodes + $::cluster_replica_nodes}]
 
-    for {set i 0} {$i < $total_instances} {incr i} {
-        wait_for_condition 1000 10 {
-            [CI $i slot_migration_active_tasks] == 0 &&
-            [CI $i slot_migration_active_trim_running] == 0
-        } else {
+    wait_for_condition 1000 10 {
+        [asm_all_instances_idle $total_instances] == 1
+    } else {
+        # Print the number of active tasks on each instance
+        for {set i 0} {$i < $total_instances} {incr i} {
             set migration_count [CI $i slot_migration_active_tasks]
             set trim_count [CI $i slot_migration_active_trim_running]
-            fail "ASM tasks did not complete on instance $i: migration_tasks=$migration_count, trim_tasks=$trim_count"
+            puts "Instance $i: migration_tasks=$migration_count, trim_tasks=$trim_count"
         }
+        fail "ASM tasks did not complete on all instances"
     }
 }
 
-proc wait_for_failover {node_id} {
-    wait_for_condition 1000 50 {
-        [string match "*master*" [R $node_id role]]
-    } else {
-        fail "Failover did not complete"
+proc wait_for_failover {node_id {failover_arg ""}} {
+    set max_attempts 5
+    for {set attempt 1} {$attempt <= $max_attempts} {incr attempt} {
+        if {$failover_arg eq ""} {
+            R $node_id cluster failover
+        } else {
+            R $node_id cluster failover $failover_arg
+        }
+
+        set completed 1
+        wait_for_condition 1000 10 {
+            [string match "*master*" [R $node_id role]]
+        } else {
+            set completed 0
+        }
+
+        if {$completed} {
+            wait_for_cluster_propagation
+            return
+        }
     }
-    wait_for_cluster_propagation
+    fail "Failover did not complete after $max_attempts attempts for node $node_id"
```

Comment:
looks like failover may fail somehow, I switched to this retry mechanism, in 10+ runs it was stable.

Reply:
i did not encounter this failure

Reply:
before you use this function to check if failover is done after trigger a failover, i think it is good enough, but now in this function, this proc  triggers failover and wait it done, but still use old proc name